src/editor/OptionsPanel.js

```diff
@@ -0,0 +1,144 @@
+            fontStyle: "italic",
+            padding: 6
+          }}
+        >
+          Multiple layers selected
+        </div>
+      </>
+    );
+  }
+
+  const selectedLayerIndex = doc.layers.findIndex(
+    ({ id }) => id === [...selection][0]
+  );
+  const selectedLayer = doc.layers[selectedLayerIndex];
+  const selectedLayerComponent = config.components.find(
+    ({ type }) => selectedLayer.component === type
+  );
+
+  // no options in selected layer component
+  if (!selectedLayerComponent.options) {
+    return (
+      <>
+        <PanelTitle style={{ marginTop: 6 }}>Options</PanelTitle>
+        <div
+          style={{
+            color: "#999",
+            fontStyle: "italic",
+            padding: 6
+          }}
+        >
+          No options
+        </div>
+      </>
+    );
+  }
+
+  const options = reconcileOptions(
+    { type: "Record", fields: selectedLayerComponent.options },
+    selectedLayer.options
+  );
+
+  const { options: displayOptions, depth: optionDepth } = resolvePath({
+    options,
+    path,
+    depth
+  });
+
+  return (
+    <>
+      <PanelTitle style={{ marginTop: 6 }}>Options</PanelTitle>
+      {/* PATH NAVBAR */}
+      {["Root", ...path].map((pathItem, i) => {
+        return (
+          <button
+            key={path.slice(i).join(".")}
+            className="path"
+            onClick={() => {
+              setPath(currPath => {
+                if (pathItem === "Root" && currPath.length === 0) {
+                  return []; // Render root level
+                }
+
+                const keyIndex = currPath.findIndex(
+                  pathName => pathName === pathItem
+                );
+                return currPath.slice(0, keyIndex + 1);
+              });
+            }}
+          >
+            {pathItem}
+            {/* Don't render the ">" for the last path element. */}
+            {path.length === i ? "" : ">"}
+            <style jsx>{`
+              .path:hover {
+                color: "pink";
+              }
+            `}</style>
+          </button>
+        );
+      })}
+
+      {/* BACK BUTTON */}
+      <button
+        onClick={() => {
+          setPath(currPath => currPath.slice(0, currPath.length - 1));
+        }}
+        // DISABLE WHEN ONLY ROOT
+        disabled={path.length === 0}
+      >
+        Back
+      </button>
+
+      {/* Options */}
+      <AbstractRenderer
+        options={displayOptions}
+        // update doc path
+        onNavigate={newPath => setPath(newPath)}
+        // used to immutably update nested object state of unknown depth
+        onChange={(newValue, rendererPath) => {
+          // rendererPath is formatted like path1.path2.path3 etc.
+          const formattedPath =
+            `doc.layers.${selectedLayerIndex}.options.` +
+            rendererPath.join(".");
+          return setState(currState => {
+            return immutable.set(currState, formattedPath, newValue);
+          });
+        }}
```

Comment:
I think there's a better way to handle these updates. Because we've got this nice recursive pattern happening, we can let the `ListRenderer` and `RecordRenderer` handle the complexity and none of the other components will need to do anything special (they wont need to provide their path when updating). I've put together a little demo of what I mean in CodeSandbox:

https://codesandbox.io/s/modern-darkness-gc6kc

In that example you can see that each component is passed an `updateNode` prop that is just an updater function. In the case where we recurse we create a new updater function for each child in created by wrapping the parent's updater function.

This will avoid the tricky problem you're using `object-path-immutable` to solve and instead we do something like this (I haven't actually tested this code, probably buggy):

```suggestion
        onChange={(updater) => {
          setState((current) => ({
            ...current,
            layers: current.layers.map((layer) => (
              layer.id === selectedLayer.id ? {
                ...layer,
                options: updater(layer.options)
              } : layer
             ))
          }));
        }}
```